enclave-runner/src/usercalls/interface.rs

```diff
@@ -13,149 +13,283 @@ use fortanix_sgx_abi::*;
+            .boxed_local()
     }
 
-    fn send(&mut self, event_set: u64, tcs: Option<Tcs>) -> UsercallResult<Result> {
-        Ok(self.0.send(event_set, tcs).to_sgx_result())
+    fn send(
+        self,
+        event_set: u64,
+        tcs: Option<Tcs>,
+    ) -> std::pin::Pin<Box<dyn Future<Output = (Self, UsercallResult<Result>)> + 'future>> {
+        async move {
+            let ret = Ok(self.0.send(event_set, tcs).to_sgx_result());
+            return (self, ret);
+        }
+            .boxed_local()
     }
 
-    fn insecure_time(&mut self) -> UsercallResult<u64> {
-        Ok(self.0.insecure_time())
+    fn insecure_time(
+        self,
+    ) -> std::pin::Pin<Box<dyn Future<Output = (Self, UsercallResult<u64>)> + 'future>> {
+        async move {
+            let ret = Ok(self.0.insecure_time());
+            return (self, ret);
+        }
+            .boxed_local()
     }
 
-    fn alloc(&mut self, size: usize, alignment: usize) -> UsercallResult<(Result, *mut u8)> {
-        Ok(self.0.alloc(size, alignment).to_sgx_result())
+    fn alloc(
+        self,
+        size: usize,
+        alignment: usize,
+    ) -> std::pin::Pin<Box<dyn Future<Output = (Self, UsercallResult<(Result, *mut u8)>)> + 'future>>
+    {
+        async move {
+            let ret = Ok(self.0.alloc(size, alignment).to_sgx_result());
+            return (self, ret);
+        }
+            .boxed_local()
     }
 
-    fn free(&mut self, ptr: *mut u8, size: usize, alignment: usize) -> UsercallResult<()> {
-        Ok(self.0.free(ptr, size, alignment).unwrap())
+    fn free(
+        self,
+        ptr: *mut u8,
+        size: usize,
+        alignment: usize,
+    ) -> std::pin::Pin<Box<dyn Future<Output = (Self, UsercallResult<()>)> + 'future>> {
+        async move {
+            let ret = Ok(self.0.free(ptr, size, alignment).unwrap());
+            return (self, ret);
+        }
+            .boxed_local()
     }
 
     fn async_queues(
-        &mut self,
+        self,
         usercall_queue: *mut FifoDescriptor<Usercall>,
         return_queue: *mut FifoDescriptor<Return>,
-    ) -> UsercallResult<Result> {
-        unsafe {
-            Ok((|| {
-                let usercall_queue = usercall_queue
-                    .as_mut()
-                    .ok_or(IoError::from(IoErrorKind::InvalidInput))?;
-                let return_queue = return_queue
-                    .as_mut()
-                    .ok_or(IoError::from(IoErrorKind::InvalidInput))?;
-                self.0.async_queues(usercall_queue, return_queue)
-            })()
-            .to_sgx_result())
+    ) -> std::pin::Pin<Box<dyn Future<Output = (Self, UsercallResult<Result>)> + 'future>> {
+        async move {
+            unsafe {
+                let ret = Ok((|| {
+                    let usercall_queue = usercall_queue
+                        .as_mut()
+                        .ok_or(IoError::from(IoErrorKind::InvalidInput));
+                    if let Err(e) = usercall_queue {
```

Comment:
you can use the ? operator instead and annotate the closure return type if needed